parser/types/field_type.go

```diff
@@ -368,12 +449,50 @@ func (ft *FieldType) StorageLength() int {
 // HasCharset indicates if a COLUMN has an associated charset. Returning false here prevents some information
 // statements(like `SHOW CREATE TABLE`) from attaching a CHARACTER SET clause to the column.
 func HasCharset(ft *FieldType) bool {
-	switch ft.Tp {
+	switch ft.tp {
 	case mysql.TypeVarchar, mysql.TypeString, mysql.TypeVarString, mysql.TypeBlob,
 		mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
-		return !mysql.HasBinaryFlag(ft.Flag)
+		return !mysql.HasBinaryFlag(ft.flag)
 	case mysql.TypeEnum, mysql.TypeSet:
 		return true
 	}
 	return false
 }
+
+// for json
+type jsonFieldType struct {
```

Comment:
Needed. Since fields are made private, json encoder/decoder will not encode/decoded these fields anymore. That said we need a custom encoder/decoder, and jsonFieldType is for that purpose...

This is a unavoidable limitation if we want to make fields private, sadly.